File: api/compare.js
import OpenAI from "openai";

const client = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

export default async function handler(req, res) {
  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed" });
  }

  try {
    const { articles } = req.body;

    if (!articles || articles.length !== 2) {
      return res.status(400).json({ error: "Exactly 2 articles required" });
    }

    const response = await client.responses.create({
      model: "gpt-4.1-mini",
      input: `
Compare these two news articles.

Return ONLY valid JSON with this shape:
{
  "article1": {
    "perspectiveTitle": "",
    "source": "",
    "coreArgument": "",
    "keyPoints": ["", "", ""]
  },
  "article2": {
    "perspectiveTitle": "",
    "source": "",
    "coreArgument": "",
    "keyPoints": ["", "", ""]
  },
  "keyDifferences": [
    {
      "category": "",
      "article1View": "",
      "article2View": ""
    }
  ]
}

Article 1:
Title: ${articles[0].title}
Source: ${articles[0].src}
Date: ${articles[0].isoTime || articles[0].date || ""}
Summary: ${articles[0].summary || ""}

Article 2:
Title: ${articles[1].title}
Source: ${articles[1].src}
Date: ${articles[1].isoTime || articles[1].date || ""}
Summary: ${articles[1].summary || ""}
      `,
    });

    const text = response.output_text;
    const json = JSON.parse(text);

    return res.status(200).json(json);
  } catch (err) {
    console.error(err);
    return res.status(500).json({ error: "Comparison failed" });
  }
}

File: api/compare.py
import json
import os

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from openai import OpenAI

app = FastAPI()


@app.post("/")
@app.post("/api/compare")
async def compare_articles(request: Request):
    try:
        body = await request.json()
        articles = body.get("articles", [])

        if len(articles) < 2:
            return JSONResponse({"error": "Two articles required."}, status_code=400)

        a1 = articles[0]
        a2 = articles[1]

        client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

        prompt = f"""
Compare these two news articles.

Article 1:
Title: {a1.get("title", "")}
Source: {a1.get("src", "")}
Summary: {a1.get("summary", "")}

Article 2:
Title: {a2.get("title", "")}
Source: {a2.get("src", "")}
Summary: {a2.get("summary", "")}

Return ONLY valid JSON with this exact structure:

{{
  "article1": {{
    "title": "...",
    "source": "...",
    "core_argument": "...",
    "key_points": ["...", "...", "..."]
  }},
  "article2": {{
    "title": "...",
    "source": "...",
    "core_argument": "...",
    "key_points": ["...", "...", "..."]
  }},
  "key_differences": [
    {{
      "label": "Framing",
      "article1": "...",
      "article2": "..."
    }},
    {{
      "label": "Tone",
      "article1": "...",
      "article2": "..."
    }},
    {{
      "label": "Emphasis",
      "article1": "...",
      "article2": "..."
    }}
  ]
}}
"""

        response = client.chat.completions.create(
            model="gpt-4.1-mini",
            response_format={"type": "json_object"},
            messages=[
                {"role": "system", "content": "You compare news articles. Return only valid JSON."},
                {"role": "user", "content": prompt},
            ],
            temperature=0.3,
        )

        comparison = json.loads(response.choices[0].message.content)
        return {"comparison": comparison}

    except Exception as e:
        return JSONResponse({"error": str(e)}, status_code=500)